File: app/src/main/java/com/spidroid/starry/adapters/UserAdapter.java
package com.spidroid.starry.adapters;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.ListAdapter;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.spidroid.starry.R;
import com.spidroid.starry.models.UserModel;

import de.hdodenhof.circleimageview.CircleImageView;

public class UserAdapter extends ListAdapter<UserModel, UserAdapter.UserViewHolder> {

    private final OnUserClickListener listener;

    public interface OnUserClickListener {
        void onUserClick(UserModel user);
    }

    public UserAdapter(OnUserClickListener listener) {
        super(DIFF_CALLBACK);
        this.listener = listener;
    }

    private static final DiffUtil.ItemCallback<UserModel> DIFF_CALLBACK =
            new DiffUtil.ItemCallback<UserModel>() {
                @Override
                public boolean areItemsTheSame(@NonNull UserModel oldItem, @NonNull UserModel newItem) {
                    return oldItem.getUserId().equals(newItem.getUserId());
                }

                @Override
                public boolean areContentsTheSame(@NonNull UserModel oldItem, @NonNull UserModel newItem) {
                    // يجب أن تتضمن هذه الدالة مقارنة لجميع الخصائص ذات الصلة التي قد تتغير
                    // للحفاظ على الأداء، قارن فقط الخصائص التي تؤثر على عرض العنصر
                    return oldItem.getDisplayName().equals(newItem.getDisplayName()) &&
                            oldItem.getUsername().equals(newItem.getUsername()) &&
                            oldItem.getProfileImageUrl().equals(newItem.getProfileImageUrl());
                    // يمكنك إضافة المزيد من المقارنات هنا إذا لزم الأمر
                }
            };

    @NonNull
    @Override
    public UserViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        // نستخدم item_user.xml الذي تم تقديمه سابقاً
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_user, parent, false);
        return new UserViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull UserViewHolder holder, int position) {
        UserModel user = getItem(position);
        holder.bind(user, listener);
    }

    static class UserViewHolder extends RecyclerView.ViewHolder {
        private final CircleImageView ivAvatar;
        private final TextView tvName;

        UserViewHolder(@NonNull View itemView) {
            super(itemView);
            ivAvatar = itemView.findViewById(R.id.iv_avatar);
            tvName = itemView.findViewById(R.id.tv_name);
        }

        void bind(UserModel user, OnUserClickListener listener) {
            tvName.setText(user.getDisplayName() != null ? user.getDisplayName() : user.getUsername());
            Glide.with(itemView.getContext())
                    .load(user.getProfileImageUrl())
                    .placeholder(R.drawable.ic_default_avatar)
                    .error(R.drawable.ic_default_avatar)
                    .into(ivAvatar);

            itemView.setOnClickListener(v -> {
                if (listener != null) {
                    listener.onUserClick(user);
                }
            });
        }
    }
}

File: app/src/main/java/com/spidroid/starry/adapters/UserAdapter.kt
package com.spidroid.starry.adapters

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.spidroid.starry.R
import com.spidroid.starry.models.UserModel
import de.hdodenhof.circleimageview.CircleImageView

class UserAdapter(private val listener: OnUserClickListener?) :
    ListAdapter<UserModel, UserAdapter.UserViewHolder?>(UserAdapter.Companion.DIFF_CALLBACK) {
    interface OnUserClickListener {
        fun onUserClick(user: UserModel?)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): UserViewHolder {
        // نستخدم item_user.xml الذي تم تقديمه سابقاً
        val view =
            LayoutInflater.from(parent.getContext()).inflate(R.layout.item_user, parent, false)
        return UserViewHolder(view)
    }

    override fun onBindViewHolder(holder: UserViewHolder, position: Int) {
        val user = getItem(position)
        holder.bind(user, listener)
    }

    internal class UserViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private val ivAvatar: CircleImageView
        private val tvName: TextView

        init {
            ivAvatar = itemView.findViewById<CircleImageView>(R.id.iv_avatar)
            tvName = itemView.findViewById<TextView>(R.id.tv_name)
        }

        fun bind(user: UserModel, listener: OnUserClickListener?) {
            tvName.setText(if (user.getDisplayName() != null) user.getDisplayName() else user.getUsername())
            Glide.with(itemView.getContext())
                .load(user.getProfileImageUrl())
                .placeholder(R.drawable.ic_default_avatar)
                .error(R.drawable.ic_default_avatar)
                .into(ivAvatar)

            itemView.setOnClickListener(View.OnClickListener { v: View? ->
                if (listener != null) {
                    listener.onUserClick(user)
                }
            })
        }
    }

    companion object {
        private val DIFF_CALLBACK: DiffUtil.ItemCallback<UserModel?> =
            object : DiffUtil.ItemCallback<UserModel?>() {
                override fun areItemsTheSame(oldItem: UserModel, newItem: UserModel): Boolean {
                    return oldItem.getUserId() == newItem.getUserId()
                }

                override fun areContentsTheSame(oldItem: UserModel, newItem: UserModel): Boolean {
                    // يجب أن تتضمن هذه الدالة مقارنة لجميع الخصائص ذات الصلة التي قد تتغير
                    // للحفاظ على الأداء، قارن فقط الخصائص التي تؤثر على عرض العنصر
                    return oldItem.getDisplayName() == newItem.getDisplayName() &&
                            oldItem.getUsername() == newItem.getUsername() &&
                            oldItem.getProfileImageUrl() == newItem.getProfileImageUrl()
                    // يمكنك إضافة المزيد من المقارنات هنا إذا لزم الأمر
                }
            }
    }
}